pkg/volume/validation/pv_validation.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/volume"
```

Comment:
As an example:

```
func ValidateNFSSource(source *api.NFSVolumeSource) .. {
  // check all valid nfs sources
}
```

Validation is an API concern - not a plugin concern.  If we expose an API, we must decide (at the time of the api) what mount options to accept.  If we are going to validate, the validation should be part of the api (if a future version of NFS adds more mount options, that's an API change).  If we don't validate, then it's not an API change.

The NFS author must specify all accepted mount options for a particular version of kubernetes, and if they want to validate them, follow the rules for introducing and deprecating mount options (we only loosen validation, we rarely tighten it).

I.e. if in 1.6 we support 3 options on NFS: "squash", "root", "uid", and validation checks those, when we later add support for "noroot", we'll allow it to be passed.  But a 1.6 cluster would not allow it unless we patched the validation.

Validation code referenced by the API needs to be completely static - the API server either decides what is valid, or checks a few options and passes the rest through, but the plugin author should be adding their validation rules into a package that only calls validation, and then importing the validation into API validation and also reusing it in the plugin (or doing more extensive validation in the plugin, and letting simpler validation happen in the API code).